compiler/src/iree/compiler/Dialect/Stream/Transforms/test/specialize_encodings.mlir

```diff
@@ -34,6 +34,101 @@ util.func public @tensor_sizeof(%d0: index, %d1: index) -> (index, index) {
 
 // -----
 
+//------------------------------------------------------------------------------
+// #iree_gpu.gpu_encoding_layout specialization tests.
+// These get serialized to the layout attributes.
+//------------------------------------------------------------------------------
+
+#map0 = affine_map<(m, n, k) -> (m, k)>
+#map1 = affine_map<(m, n, k) -> (k, n)>
+#map2 = affine_map<(m, n, k) -> (m, n)>
+#executable_target_rocm_hsaco_fb = #hal.executable.target<"rocm", "rocm-hsaco-fb",
+  {
+    abi = "hip",
+    encoding = #iree_gpu.gpu_encoding_layout<>,
+    iree.gpu.target = #iree_gpu.target<arch = "gfx942",
+                                       features = "",
+                                       wgp = <compute = fp32,
+                                              storage =  b32,
+                                              subgroup =  none,
+                                              dot =  none,
+                                              mma = [<MFMA_F32_16x16x4_F32>],
+                                              subgroup_size_choices = [64],
+                                              max_workgroup_sizes = [1024, 1024, 1024],
+                                              max_thread_count_per_workgroup = 1024,
+                                              max_workgroup_memory_bytes = 65536,
+                                              max_workgroup_counts = [2147483647, 2147483647, 2147483647],
+                                              max_load_instruction_bits = 128,
+                                              simds_per_wgp = 4,
+                                              vgpr_space_bits = 16384>>
+  }>
+#device_target_local_0_ = #hal.device.target<"local", {ordinal = 0 : index}, [#executable_target_rocm_hsaco_fb]> : !hal.device
+#encoding = #iree_encoding.encoding<operand_index = 0 : index, op_type =  matmul, element_types = [f32, f32, f32], user_indexing_maps = [#map0, #map1, #map2]>
+
+util.global private @device_a = #device_target_local_0_
+util.func public @gpu_with_encoding_layout(%d0: index, %d1: index) -> index {
+  %size0 = stream.tensor.sizeof on(#hal.device.affinity<@device_a>) tensor<?x?xf32, #encoding>{%d0, %d1} : index
+  util.return %size0 : index
+}
+// CHECK:       #[[$ENCODING:.+]] = #iree_encoding.encoding
+// CHECK-SAME:    #iree_gpu.gpu_encoding_layout
+// CHECK-SAME:    encoding_info = {innerDimsPos = [{{.+}}], innerTileSizes = [{{.+}}], outerDimsPerm = [{{.+}}]}
+// CHECK-LABEL: util.func public @gpu_with_encoding_layout
+// CHECK:         %[[RES:.+]] = stream.tensor.sizeof {{.+}} tensor<?x?xf32, #[[$ENCODING]]>
+// CHECK:         return %[[RES]]
+
+// -----
+
+// A pre-existing encoding in the target attribute should be materialized in the encoding attribute
```

Comment:
Is this for testing the change of `getEncodingInfoForMatmul`? If so, I think it is better to add a related test to [codegen tests](https://github.com/iree-org/iree/blob/main/compiler/src/iree/compiler/Codegen/Common/test/gpu_materialize_encoding_gfx942.mlir) instead. It keeps the stream transformation tests cleaner because it isolates target-specific logic.

Reply:
No, this tests https://github.com/iree-org/iree/pull/20027/files#diff-4ffd17e19dafacc4ed2d618dbabbda98b906b9514d98884b1afcc6e1ec5b52eeR325. I will add an additional check for the `getEncodingInfoForMatmul` change to the codegen tests then.

Reply:
The link seems broken to me, I can't really see which code you're refering. Anyway, the pre-resolved encoding issue should be addressed by my other PR, see the other comment.

Reply:
This test is a different type of pre-resolved encoding and it tests the following logic being added in this PR, so we should have a test for it:

````
// If the layout is already resolved, use it directly.
if (config) {
  if (std::optional<NamedAttribute> namedAttr =
          config.getNamed(kEncodingInfoAttrName)) {
    std::optional<MaterializeEncodingInfo> info =
        Codegen::deserializeEncodingInfo(
            cast<DictionaryAttr>(namedAttr->getValue()));
    assert(info && "encoding_info is invalid");
    return info.value();
  }
}
```

https://github.com/iree-org/iree/pull/20027/files#diff-4ffd17e19dafacc4ed2d618dbabbda98b906b9514d98884b1afcc6e1ec5b52eeR311

Reply:
Thanks for pointing out the path through `MaterializeEncoding`. I tried to move the test there, but actually, I noticed that this piece of logic I referenced above wasn't hit correctly in that flow, so the test failed. I had to add the following logic to correctly move the encoding and then it works: https://github.com/iree-org/iree/pull/20027/files#diff-faf77ef5ad3170cb57799accef7d5993f96a72aab22468ded53ce95799181617R84. As the flows through `MaterializeEncoding` and `SpecializeEncodings` are quite different, I think both tests are warranted to ensure correctness.

Reply:
And I think there is likely something going wrong for the CPU flow through `MaterializeEncoding` as well.

Reply:
There is no other test for this piece of logic in `getEncodingInfo` through `SpecializeEncodings` and I think the `MaterializeEncoding` issue I encountered above shows that the flow into this function is quite different and we should keep both to test this well.